### PR TITLE
IPAM API: create associated Services for IP resources

### PR DESCRIPTION
**Pull request by fra98**
# Description

Added support for creating a Service (and EndpointSlice) associated to an IP resource.

Ref #1990

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [x] locally

### CODE REVIEW

**Comment by fra98** [comment]
[Controllori della famiglia IPAM](https://trello.com/c/NqmigHgF/2-controllori-della-famiglia-ipam)

File: apis/ipam/v1alpha1/ip_types.go [hunk 1]
```diff
@@ -33,10 +34,20 @@ var (
 	IPGroupResource = schema.GroupResource{Group: GroupVersion.Group, Resource: IPResource}
 )
 
+// ServiceTemplate contains the template to create the associated service (and endpointslice) for the IP endopoint.
+type ServiceTemplate struct {
+	// Template Spec of the Service.
```
Comment (aleoli) [inline apis/ipam/v1alpha1/ip_types.go:39]: I would add also metadata here to allow adding labels and annotations

File: pkg/virtualKubelet/reflection/exposition/endpointslice.go [hunk 2]
```diff
@@ -180,6 +184,17 @@ func (ner *NamespacedEndpointSliceReflector) Handle(ctx context.Context, name st
...
 	}
 
+	// If the local endpointslice has the "skip unmap ip" annotation, then we do not have to unmap the addresses as already
+	// performed by other entities. We store in a cache if the endpointslice has the annotation or not, so that we have that
+	// information even when the local endpointslice is deleted (and therefore we can't check the existence of the annotation).
+	if local.GetAnnotations() != nil {
+		_, hasSkipUnmapAnnot := local.GetAnnotations()[consts.VKSkipUnmapIPAnnotationKey]
+		if hasSkipUnmapAnnot {
+			// the map contains only the keys of the endpoinstslices with the annotation, the values are not used.
+			ner.epsSkipUnmap.Store(name, nil)
```
Comment (aleoli) [inline pkg/virtualKubelet/reflection/exposition/endpointslice.go:194]: You should remove it if present and the annotation is no more set

**Review by aleoli — APPROVED** [review]

File: apis/ipam/v1alpha1/ip_types.go [hunk 3]
```diff
@@ -33,10 +34,22 @@ var (
 	IPGroupResource = schema.GroupResource{Group: GroupVersion.Group, Resource: IPResource}
 )
 
+// ServiceTemplate contains the template to create the associated service (and endpointslice) for the IP endopoint.
+type ServiceTemplate struct {
+	// Metadata of the Service.
+	Metadata metav1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (aleoli) [inline apis/ipam/v1alpha1/ip_types.go:40]: Make sure that if it is not defined in the yaml, it works fine.

**Review by aleoli — APPROVED** [review]

**Pull request merged**